### DISCUSSION TITLE
howto fall back to regular tkinter when non supported OS encountered?

### DISCUSSION THREAD

**Discussion by Free-Radical** [TomSchimansky/CustomTkinter#541] [Q&A] [answered]
Hi, I'm writing a cross platform app that runs on Windows, Mac and Linux. In Linux,  CustomTkinter ends up looking really bad, as expected. 

What would be an elegant way to handle "fall back" to regular tkinter or similar, when a non-supported OS is encountered?

**Comment by Akascape** [comment]
@Free-Radical In that case you have to make two separate programs (one in tkinter and another in customtkinter).
Then you can use this to identify the OS: 
```python
import sys
if sys.platform.startswith("linux"):
   Normal_Tkinter() ##Link this to your normal tkinter app class/function.
```
You can also combine both programs in one file.

**Reply by Free-Radical** [reply]
My apologies but I'm a noob: how do I "link" this to my tkinter function?
I'm using the following type of calls (not using OOP)

`root_window = customtkinter.CTk()`
`PREFS_WDW = customtkinter.CTkToplevel(root_window)`

**Or are you implying writing 2 separate series of commands to call normal tkinter and CustomTkinter as required**

**Comment by RutgerLit** [comment]
> In Linux, CustomTkinter ends up looking really bad, as expected.

Could you tell me why this is expected? I was thinking about making a Linux version as well but this might stop me from doing that.

**Reply by Free-Radical** [reply]
I've put an image below how the official sample code looks, but I figured it wouldn't look good since it was custom designed to look good for Windows/Mac and not Linux?

**Comment by TomSchimansky** [comment]
What is your problem on Linux? Linux is supported by CustomTkinter. I can not test it on other distributions than Ubuntu and Kali Linux, but for me it always worked.

**Comment by Free-Radical** [comment]
Good to hear, hopefully i'm doing something wrong (which is very possible , since I'm a noob at this), here's how a window looks using your code in Linux Mint 21 Cinnamon (looks fine in Windows/Mac)

![tmp](https://user-images.githubusercontent.com/753154/197238943-8b95aad9-f94d-4ccf-8c5c-4f31f47fa938.png)

**Reply by TomSchimansky** [reply]
Are you using the latest version? Try upgrading with pip3 install customtkinter --upgrade. The were versions were this happened on Linux.

**Reply by TomSchimansky** [reply]
Also look here: https://github.com/TomSchimansky/CustomTkinter/issues/206

**Reply by TomSchimansky** [reply]
Or here: https://github.com/TomSchimansky/CustomTkinter/issues/300

**Reply by Free-Radical** [reply]
Nope pip3 says up to date-
```
$  pip3 install customtkinter --upgrade
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: customtkinter in ./.local/lib/python3.10/site-packages (4.6.3)
Requirement already satisfied: darkdetect in ./.local/lib/python3.10/site-packages (from customtkinter) (0.7.1)
```

python version Python 3.10.6
Let me know how i can help, I can spin up an Ubunu as well if u need.

**Reply by TomSchimansky** [reply]
Do you have a folder `.fonts/` in your home directory? And are there font files in there? There must be a file `CustomTkinter_shapes_font.otf` in order for the rendering to work.

**Reply by TomSchimansky** [reply]
Maybe you need to update your font cache, I found this command `fc-cache -fv` but for me this was never necessary.

**Reply by TomSchimansky** [reply]
Can you also show me what this prints for you.
```python
import sys
print(sys.platform)
```

**Reply by Free-Radical** [reply]
```
Python 3.10.6 (main, Oct  7 2022, 20:19:58) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> import sys
>>> print(sys.platform)
linux
```

**Reply by Free-Radical** [reply]
> Do you have a folder `.fonts/` in your home directory? And are there font files in there? There must be a file `CustomTkinter_shapes_font.otf` in order for the rendering to work.

YES file exists

**Reply by Free-Radical** [reply]
> Maybe you need to update your font cache, I found this command `fc-cache -fv` but for me this was never necessary.

executed command , it succeeded.
fired up app again (without reboot ,etc) SAME RESULT

**Reply by Free-Radical** [reply]
same results after reboot as well @TomSchimansky

**Comment by Akascape** [comment, marked as answer]
@Free-Radical

>I'm using the following type of calls (not using OOP)
>Or are you implying writing 2 separate series of commands to call normal tkinter and CustomTkinter as required

### Here is an Example 
```python 
import tkinter
import customtkinter
import sys
import tkinter.ttk as ttk

customtkinter.set_appearance_mode("dark")
customtkinter.set_default_color_theme("Light")  

def CTK_App():
    
    app = customtkinter.CTk()
    app.geometry("400x540")
    app.title("CustomTkinter simple_example.py")


    def button_function():
        print("Button click", label_1.text_label.cget("text"))


    def slider_function(value):
        progressbar_1.set(value)


    def check_box_function():
        print("checkbox_1:", checkbox_1.get())


    y_padding = 13

    frame_1 = customtkinter.CTkFrame(master=app, corner_radius=15)
    frame_1.pack(pady=20, padx=60, fill="both", expand=True)

    label_1 = customtkinter.CTkLabel(master=frame_1,text_font="caltons", justify=tkinter.LEFT)
    label_1.pack(pady=y_padding, padx=10)

    progressbar_1 = customtkinter.CTkProgressBar(master=frame_1)
    progressbar_1.pack(pady=y_padding, padx=10)

    button_1 = customtkinter.CTkButton(master=frame_1, corner_radius=8, command=button_function)
    button_1.pack(pady=y_padding, padx=10)

    slider_1 = customtkinter.CTkSlider(master=frame_1, command=slider_function, from_=0, to=1)
    slider_1.pack(pady=y_padding, padx=10)
    slider_1.set(0.5)

    entry_1 = customtkinter.CTkEntry(master=frame_1, placeholder_text="CTkEntry")
    entry_1.pack(pady=y_padding, padx=10)

    checkbox_1 = customtkinter.CTkCheckBox(master=frame_1, command=check_box_function)
    checkbox_1.pack(pady=y_padding, padx=10)

    radiobutton_var = tkinter.IntVar(value=1)

    radiobutton_1 = customtkinter.CTkRadioButton(master=frame_1, variable=radiobutton_var, value=1)
    radiobutton_1.pack(pady=y_padding, padx=10)

    radiobutton_2 = customtkinter.CTkRadioButton(master=frame_1, variable=radiobutton_var, value=2)
    radiobutton_2.pack(pady=y_padding, padx=10)

    s_var = tkinter.StringVar(value="on")

    switch_1 = customtkinter.CTkSwitch(master=frame_1)
    switch_1.pack(pady=y_padding, padx=10)

    optionmenu_1 = customtkinter.CTkComboBox(master=frame_1, values=["option 1", "option 2", "number 42","option 1", "option 2", "number 42","option 1", "option 2", "number 42"])
    optionmenu_1.pack(pady=y_padding, padx=10)

    app.mainloop()

def Normal_Tkinter():
    
    app = tkinter.Tk()
    app.geometry("400x400")
    app.title("Standard Tkinter Test")


    def button_function():
        print("button pressed")


    def slider_function(value):
        progressbar_1["value"] = value


    s = ttk.Style()
    s.configure("TRadiobutton", fg="red")

    y_padding = 6

    frame_1 = tkinter.Frame(master=app, width=300, height=260, bg="lightgray")
    frame_1.pack(padx=60, pady=20, fill="both", expand=True)

    label_1 = tkinter.Label(master=frame_1, text="Label", bg="lightgray")
    label_1.pack(pady=y_padding, padx=10)

    progressbar_1 = ttk.Progressbar(master=frame_1, style='black.Horizontal.TProgressbar', length=150)
    progressbar_1.pack(pady=y_padding, padx=10)
    progressbar_1["value"] = 50

    button_1 = tkinter.Button(master=frame_1, command=button_function, text="Button", highlightbackground="lightgray")
    button_1.pack(pady=y_padding, padx=10)

    slider_1 = tkinter.Scale(master=frame_1, command=slider_function, orient="horizontal", bg="lightgray", length=150)
    slider_1.pack(pady=y_padding, padx=10)

    entry_1 = tkinter.Entry(master=frame_1, highlightbackground="lightgray", width=10)
    entry_1.pack(pady=y_padding, padx=10)

    checkbox_1 = tkinter.Checkbutton(master=frame_1, bg=frame_1.cget("bg"), text="CheckButton")
    checkbox_1.pack(pady=y_padding, padx=10)

    radiobutton_var = tkinter.IntVar()

    radiobutton_1 = ttk.Radiobutton(master=frame_1, variable=radiobutton_var, value=1, text="Radiobutton")
    radiobutton_1.pack(pady=y_padding, padx=10)

    radiobutton_2 = ttk.Radiobutton(master=frame_1, variable=radiobutton_var, value=2, text="Radiobutton")
    radiobutton_2.pack(pady=y_padding, padx=10)

    combobox_1 = ttk.Combobox(master=frame_1, values=["option 1", "option 2", "number 42","option 1", "option 2", "number 42","option 1", "option 2", "number 42"])
    combobox_1.pack(pady=y_padding, padx=10)

    app.mainloop()
    
if __name__ == "__main__":
    
    # Identify the OS
    if sys.platform.startswith("linux"):
        Normal_Tkinter() # This will link it to the normal tkinter class app
    else:
        CTK_App()
```

**Reply by avalon60** [reply]
I do most of my personal project work on Linux Mint. I have never encountered that kind of problem with CustomTkinter.

**Reply by Free-Radical** [reply]
@Akascape  Thank you!

**Reply by Free-Radical** [reply]
> I do most of my personal project work on Linux Mint. I have never encountered that kind of problem with CustomTkinter.

@avalon60  do you mind sharing your system environment, kernel, python,etc to compare?
